src/cargo/core/profiles.rs

```diff
@@ -766,7 +766,17 @@ pub struct UnitFor {
     /// A target for `build.rs` or any of its dependencies, or a proc-macro or
     /// any of its dependencies. This enables `build-override` profiles for
     /// these targets.
-    build: bool,
+    host: bool,
+    /// A target for a build dependency (or any of its dependencies). This is
+    /// used for computing features of build dependencies independently of
+    /// other dependency kinds.
+    ///
+    /// The subtle difference between this and `host` is that the build script
+    /// for a non-host package sets this to `false` because it wants the
+    /// features of the non-host package (whereas `host` is true because the
+    /// build script is being built for the host). `build_dep` becomes `true`
+    /// for build-dependencies, or any of their dependencies.
+    build_dep: bool,
```

Comment:
I'm thinking about this for a bit now. I understand (I think?) all the words in this comment, but I'm having trouble piecing together a situation about where this becomes relevant. Can you give a small example about where this subtle distinction is necessary?

Additionally, since `host=false, build_dep=true` probably doesn't make much sense, would it make more sense for this to be some sort of tri-value thing like:

```rust
enum Kind {
    Target,
    Host { build_dep: bool },
}
```

Reply:
(please don't use `Kind` as I'm sure you're already thinking when reading that)

Reply:
I added an example.

I'm reluctant to add an enum to represent the 3 states. All the code is oriented around simple booleans, and I think it would make it harder to follow if they were unified into one field. I usually agree that it is good to prevent impossible scenarios, but in this case I think it might be harder to understand.

Reply:
Ok the example looks good to me, thanks! I'm also fine deferring to your thoughts about booleans! 

One final question here, though. Technically we have two sets of fetures to deal with, one when the build script is compiled (the `--feature` flags) and one when it's executed (`CARGO_FEATURE_*`). Given that `is_custom_build` is conflating both executing and compiling the build script, I think we may want to be a little more nuanced/precise? I would expect that the compilation of the build script would use the `host: true` and `build_dep: false` set of features, but the execution of the build script would use `host: false` and `build_dep: false` as well, right? (ok maybe this doesn't have to do with `build_dep`?)

Also to clarify, in the example you listed here, `shared_dep build.rs` is only compiled once regardless of feature settings, right? And then it's executed twice maybe depending on feature settings/targets?

Reply:
Hm, so it is working correctly, but you do point out that `host` isn't accurate for the RunCustomBuild unit.  `host` only matters for the profile, and the RunCustomBuild unit goes through a separate code path for computing the profile.  That is [here](https://github.com/rust-lang/cargo/blob/369991b9a09c9f49deebe81fa05d6e104c59ea11/src/cargo/core/compiler/unit_dependencies.rs#L496-L501) where it manually fetches the profile (ignoring unit_for).   All other units go through [this path](https://github.com/rust-lang/cargo/blob/369991b9a09c9f49deebe81fa05d6e104c59ea11/src/cargo/core/compiler/unit_dependencies.rs#L544-L549) where it inspects unit_for to determine the profile.  However, it is necessary for `host` to be `true` for RunCustomBuild so that all dependencies below it are marked as "host:true".  Added some comments explaining this.

> `shared_dep build.rs` is only compiled once regardless of feature settings, right? And then it's executed twice maybe depending on feature settings/targets?

It depends.  With `-Zfeatures=build_dep`, and the features listed are different between the normal and build dep, then it will get built twice.  The reason for this is that many scripts out there use `#[cfg]` and `cfg!` to detect features, instead of inspecting "CARGO_FEATURE_*".   Only building it once would break a large number of build scripts.  I don't think it would be feasible to avoid that.

Reply:
Ah ok that makes sense, I know I've "used and abused" that `CARGO_FEATURE_*` is the "same" as `#[cfg]` in the past myself, so having a special case for this makes sense.